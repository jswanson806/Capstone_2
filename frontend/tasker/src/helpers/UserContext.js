import React, { createContext, useState } from 'react';

// Create the context
const UserContext = createContext();

// Create a provider for the context
const UserProvider = ({ children }) => {
  const [user, setUser] = useState(localStorage.getItem('user') || '');

  // Function to update the user in the state and localStorage
  const updateUser = (newUser) => {
    setUser(newUser);
    localStorage.setItem('user', newUser);
  };

  return (
    <UserContext.Provider value={{ user, updateUser }}>
      {children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
