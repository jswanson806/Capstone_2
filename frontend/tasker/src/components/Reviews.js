import React, {useContext, useEffect, useState} from 'react';
import ReviewCard from './ReviewCard.js';
import TaskerApi from '../api.js';
import { UserContext } from '../helpers/UserContext.js';

const Reviews = () => {

    const { user } = useContext(UserContext);

    const [reviews, setReviews] = useState([]);
    const [reviewCards, setReviewCards] = useState([]);
    const [toggleReviews, setToggleReviews] = useState(false);

    useEffect(() => {
        fetchAndSetReviews();
    }, []);


    useEffect(() => {

        if(reviews.length){
            setToggleReviews(true);
            createReviewCards();
        }
        
    }, [reviews]);


    const fetchAndSetReviews = async () => {
        try {
            const user_id = JSON.parse(user).id;
            const res = await TaskerApi.getAllReviewsForUser(user_id);
            const { reviews } = res;
            setReviews(reviews);
        } catch {
            return;
        };
    };


    const createReviewCards = () => {
        setReviewCards(reviews.map((review) => {
            return <ReviewCard review={review} key={review.id}/>
        }));
    };

    return (

        <div>
            <h1>Reviews</h1>
            <div>
                {toggleReviews && ([reviewCards])}
                {!toggleReviews && (<p>No reviews yet!</p>)}
            </div>
        </div>

    )
}

export default Reviews;