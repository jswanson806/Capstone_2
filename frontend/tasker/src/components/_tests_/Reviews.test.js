import React from 'react';
import { render, act } from '@testing-library/react';
import Reviews from '../Reviews.js';
import { UserContext } from '../../helpers/UserContext.js';
import TaskerApi from '../../api.js';

const reviews = {reviews: [
    {   
        id: 1, 
        title: 'test review 1', 
        body: 'test review 1 body', 
        stars: 4, 
        reviewed_by: 2, 
        reviewed_for: 1
    }, 
    {
        id: 2, 
        title: 'test review 2', 
        body: 'test review 2 body', 
        stars: 3, 
        reviewed_by: 3, 
        reviewed_for: 1
    }
]};

const userValue = { user: '{"id": 1, "email": "test@email.com", "isWorker": true}' };

const getAllReviewsForUser = jest.spyOn(TaskerApi, 'getAllReviewsForUser');

describe("smoke and snapshot tests", () => {

    beforeEach(() => {
        getAllReviewsForUser.mockResolvedValueOnce(reviews)
    });

    afterEach(() => {
        jest.clearAllMocks();
    })
    
    test("should render without crashing", async () => {
        await act(async () => {render(
            <UserContext.Provider value={userValue} >
                <Reviews />
            </UserContext.Provider>
        )})
    }); 
    
    test("should match the snapshot", async () => {

        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <UserContext.Provider value={userValue} >
                <Reviews />
            </UserContext.Provider>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    });

});


describe("useEffect hooks", () => {

    beforeEach(() => {
        getAllReviewsForUser.mockResolvedValueOnce(reviews)
    });

    afterEach(() => {
        jest.clearAllMocks();
    })
    
    test("should call api to get reviews for user", async () => {
        await act(async () => {render(
            <UserContext.Provider value={userValue} >
                <Reviews />
            </UserContext.Provider>
        )})

        expect(getAllReviewsForUser).toHaveBeenCalledTimes(1);
        expect(getAllReviewsForUser).toHaveBeenCalledWith(1);
    });
    
    
    test("should call createReviewCards function - indirect test", async () => {

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue} >
                <Reviews />
            </UserContext.Provider>
        )
        });
        
        const { container } = renderResult;
            
        expect(container).toHaveTextContent('test review 1');
        expect(container).toHaveTextContent('test review 2');

    });

});

describe('conditional rendering of reviews', () => {

    test("should show message if no reviews available", async () => {
        getAllReviewsForUser.mockResolvedValueOnce(null);
        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue} >
                <Reviews />
            </UserContext.Provider>
        )
        });
        
        const { container } = renderResult;
            
        expect(container).not.toHaveTextContent('test review 1');
        expect(container).not.toHaveTextContent('test review 2');
        expect(container).toHaveTextContent('No reviews yet!');

    });
    
});