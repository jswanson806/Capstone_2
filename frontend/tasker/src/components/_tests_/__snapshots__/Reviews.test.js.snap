// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Reviews
    </h1>
    <div>
      <div>
        <h1>
          test review 1
        </h1>
        <h3>
          4
        </h3>
        <p>
          test review 1 body
        </p>
      </div>
      <div>
        <h1>
          test review 2
        </h1>
        <h3>
          3
        </h3>
        <p>
          test review 2 body
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
